In [1]:
import sqlalchemy

In [2]:
sqlalchemy.create_engine('sqlite:///factbook.db')

Engine(sqlite:///factbook.db)

In [3]:
%%capture
%load_ext sql

In [4]:
%sql sqlite:///factbook.db

In [5]:
%%sql
SELECT * FROM sqlite_master WHERE type='table';

 * sqlite:///factbook.db
Done.


[('table', 'sqlite_sequence', 'sqlite_sequence', 3, 'CREATE TABLE sqlite_sequence(name,seq)'),
 ('table', 'facts', 'facts', 47, 'CREATE TABLE "facts" ("id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, "code" varchar(255) NOT NULL, "name" varchar(255) NOT NULL, "area" integer, "a ... (4 characters truncated) ... land" integer, "area_water" integer, "population" integer, "population_growth" float, "birth_rate" float, "death_rate" float, "migration_rate" float)')]

In [6]:
%%sql
SELECT * FROM facts
LIMIT 5;

 * sqlite:///factbook.db
Done.


[(1, 'af', 'Afghanistan', 652230, 652230, 0, 32564342, 2.32, 38.57, 13.89, 1.51),
 (2, 'al', 'Albania', 28748, 27398, 1350, 3029278, 0.3, 12.92, 6.58, 3.3),
 (3, 'ag', 'Algeria', 2381741, 2381741, 0, 39542166, 1.84, 23.67, 4.31, 0.92),
 (4, 'an', 'Andorra', 468, 468, 0, 85580, 0.12, 8.13, 6.96, 0.0),
 (5, 'ao', 'Angola', 1246700, 1246700, 0, 19625353, 2.78, 38.78, 11.49, 0.46)]

Let's start by calculating some summary statistics and look for any outlier countries.


In [7]:
%%sql
SELECT MIN(population), MAX(population), MIN(population_growth), MAX(population_growth)
FROM facts;

 * sqlite:///factbook.db
Done.


[(0, 7256490011, 0.0, 4.02)]

A few things stick out from the summary statistics in the last screen:

there's a country with a population of 0
there's a country with a population of 7256490011 (or more than 7.2 billion people)
Let's use subqueries to zoom in on just these countries without using the specific values.



In [9]:
%%sql
SELECT name FROM facts
WHERE population = (SELECT MIN(population) FROM facts)

 * sqlite:///factbook.db
Done.


[('Antarctica',)]

In [10]:
%%sql
SELECT name FROM facts
WHERE population = (SELECT MAX(population) FROM facts)

 * sqlite:///factbook.db
Done.


[('World',)]

Let's continue by calculating some averages.


In [12]:
%%sql
SELECT AVG(population) AS avg_population, AVG(area) AS avg_area
FROM facts

 * sqlite:///factbook.db
Done.


[(62094928.32231405, 555093.546184739)]

We'll identify countries that have:
- Above average values for population.
- Below average values for area.

In [13]:
%%sql
SELECT name FROM facts 
WHERE population > (SELECT AVG(population) FROM facts)

 * sqlite:///factbook.db
Done.


[('Bangladesh',),
 ('Brazil',),
 ('China',),
 ('Congo, Democratic Republic of the',),
 ('Egypt',),
 ('Ethiopia',),
 ('France',),
 ('Germany',),
 ('India',),
 ('Indonesia',),
 ('Iran',),
 ('Japan',),
 ('Mexico',),
 ('Nigeria',),
 ('Pakistan',),
 ('Philippines',),
 ('Russia',),
 ('Thailand',),
 ('Turkey',),
 ('United Kingdom',),
 ('United States',),
 ('Vietnam',),
 ('European Union',),
 ('World',)]

In [14]:
%%sql
SELECT name FROM facts 
WHERE area < (SELECT AVG(area) FROM facts)

 * sqlite:///factbook.db
Done.


[('Albania',),
 ('Andorra',),
 ('Antigua and Barbuda',),
 ('Armenia',),
 ('Austria',),
 ('Azerbaijan',),
 ('Bahamas, The',),
 ('Bahrain',),
 ('Bangladesh',),
 ('Barbados',),
 ('Belarus',),
 ('Belgium',),
 ('Belize',),
 ('Benin',),
 ('Bhutan',),
 ('Bosnia and Herzegovina',),
 ('Brunei',),
 ('Bulgaria',),
 ('Burkina Faso',),
 ('Burundi',),
 ('Cambodia',),
 ('Cameroon',),
 ('Cabo Verde',),
 ('Comoros',),
 ('Congo, Republic of the',),
 ('Costa Rica',),
 ("Cote d'Ivoire",),
 ('Croatia',),
 ('Cuba',),
 ('Cyprus',),
 ('Czech Republic',),
 ('Denmark',),
 ('Djibouti',),
 ('Dominica',),
 ('Dominican Republic',),
 ('Ecuador',),
 ('El Salvador',),
 ('Equatorial Guinea',),
 ('Eritrea',),
 ('Estonia',),
 ('Fiji',),
 ('Finland',),
 ('Gabon',),
 ('Gambia, The',),
 ('Georgia',),
 ('Germany',),
 ('Ghana',),
 ('Greece',),
 ('Grenada',),
 ('Guatemala',),
 ('Guinea',),
 ('Guinea-Bissau',),
 ('Guyana',),
 ('Haiti',),
 ('Honduras',),
 ('Hungary',),
 ('Iceland',),
 ('Iraq',),
 ('Ireland',),
 ('Israel',),
 ('Italy',),
 ('Jamaica',),
 ('Japan',),
 ('Jordan',),
 ('Kiribati',),
 ('Korea, North',),
 ('Korea, South',),
 ('Kosovo',),
 ('Kuwait',),
 ('Kyrgyzstan',),
 ('Laos',),
 ('Latvia',),
 ('Lebanon',),
 ('Lesotho',),
 ('Liberia',),
 ('Liechtenstein',),
 ('Lithuania',),
 ('Luxembourg',),
 ('Macedonia',),
 ('Malawi',),
 ('Malaysia',),
 ('Maldives',),
 ('Malta',),
 ('Marshall Islands',),
 ('Mauritius',),
 ('Micronesia, Federated States of',),
 ('Moldova',),
 ('Monaco',),
 ('Montenegro',),
 ('Morocco',),
 ('Nauru',),
 ('Nepal',),
 ('Netherlands',),
 ('New Zealand',),
 ('Nicaragua',),
 ('Norway',),
 ('Oman',),
 ('Palau',),
 ('Panama',),
 ('Papua New Guinea',),
 ('Paraguay',),
 ('Philippines',),
 ('Poland',),
 ('Portugal',),
 ('Qatar',),
 ('Romania',),
 ('Rwanda',),
 ('Saint Kitts and Nevis',),
 ('Saint Lucia',),
 ('Saint Vincent and the Grenadines',),
 ('Samoa',),
 ('San Marino',),
 ('Sao Tome and Principe',),
 ('Senegal',),
 ('Serbia',),
 ('Seychelles',),
 ('Sierra Leone',),
 ('Singapore',),
 ('Slovakia',),
 ('Slovenia',),
 ('Solomon Islands',),
 ('Spain',),
 ('Sri Lanka',),
 ('Suriname',),
 ('Swaziland',),
 ('Sweden',),
 ('Switzerland',),
 ('Syria',),
 ('Tajikistan',),
 ('Thailand',),
 ('Timor-Leste',),
 ('Togo',),
 ('Tonga',),
 ('Trinidad and Tobago',),
 ('Tunisia',),
 ('Turkmenistan',),
 ('Tuvalu',),
 ('Uganda',),
 ('United Arab Emirates',),
 ('United Kingdom',),
 ('Uruguay',),
 ('Uzbekistan',),
 ('Vanuatu',),
 ('Holy See (Vatican City)',),
 ('Vietnam',),
 ('Yemen',),
 ('Zimbabwe',),
 ('Taiwan',),
 ('Ashmore and Cartier Islands',),
 ('Christmas Island',),
 ('Cocos (Keeling) Islands',),
 ('Coral Sea Islands',),
 ('Heard Island and McDonald Islands',),
 ('Norfolk Island',),
 ('Hong Kong',),
 ('Macau',),
 ('Faroe Islands',),
 ('Clipperton Island',),
 ('French Polynesia',),
 ('New Caledonia',),
 ('Saint Martin',),
 ('Saint Pierre and Miquelon',),
 ('Wallis and Futuna',),
 ('Aruba',),
 ('Curacao',),
 ('Sint Maarten',),
 ('Cook Islands',),
 ('Niue',),
 ('Tokelau',),
 ('Bouvet Island',),
 ('Jan Mayen',),
 ('Svalbard',),
 ('Akrotiri',),
 ('Anguilla',),
 ('Bermuda',),
 ('British Indian Ocean Territory',),
 ('British Virgin Islands',),
 ('Cayman Islands',),
 ('Dhekelia',),
 ('Falkland Islands (Islas Malvinas)',),
 ('Gibraltar',),
 ('Guernsey',),
 ('Jersey',),
 ('Isle of Man',),
 ('Montserrat',),
 ('Pitcairn Islands',),
 ('Saint Helena, Ascension, and Tristan da Cunha',),
 ('South Georgia and South Sandwich Islands',),
 ('Turks and Caicos Islands',),
 ('American Samoa',),
 ('Guam',),
 ('Navassa Island',),
 ('Northern Mariana Islands',),
 ('Puerto Rico',),
 ('Virgin Islands',),
 ('Wake Island',),
 ('Gaza Strip',),
 ('Paracel Islands',),
 ('Spratly Islands',),
 ('West Bank',),
 ('Western Sahara',)]

Which countries have the highest ratios of water to land? 

In [26]:
%%sql

SELECT name, ROUND(CAST(area_water AS float) / CAST(area_land AS float), 3) AS water_land_ratio
FROM facts
ORDER BY water_land_ratio DESC
LIMIT 10

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory', 905.667),
 ('Virgin Islands', 4.52),
 ('Puerto Rico', 0.555),
 ('Bahamas, The', 0.387),
 ('Guinea-Bissau', 0.285),
 ('Malawi', 0.259),
 ('Netherlands', 0.226),
 ('Uganda', 0.223),
 ('Eritrea', 0.164),
 ('Liberia', 0.156)]

Which countries have more water than land?

In [29]:
%%sql
SELECT name FROM facts
WHERE area_water > area_land

 * sqlite:///factbook.db
Done.


[('British Indian Ocean Territory',), ('Virgin Islands',)]

Which countries will add the most people to their population next year?

In [32]:
%%sql
SELECT name, population_growth
FROM facts
ORDER BY population_growth DESC
LIMIT 10 

 * sqlite:///factbook.db
Done.


[('South Sudan', 4.02),
 ('Malawi', 3.32),
 ('Burundi', 3.28),
 ('Niger', 3.25),
 ('Uganda', 3.24),
 ('Qatar', 3.07),
 ('Burkina Faso', 3.03),
 ('Mali', 2.98),
 ('Cook Islands', 2.95),
 ('Iraq', 2.93)]

Which countries have a higher death rate than birth rate?

In [36]:
%%sql 
SELECT name, death_rate - birth_rate AS difference FROM facts
WHERE death_rate > birth_rate
ORDER BY difference DESC


 * sqlite:///factbook.db
Done.


[('Bulgaria', 5.52),
 ('Serbia', 4.58),
 ('Latvia', 4.3100000000000005),
 ('Lithuania', 4.17),
 ('Ukraine', 3.74),
 ('Hungary', 3.5700000000000003),
 ('Germany', 2.9499999999999993),
 ('Slovenia', 2.9499999999999993),
 ('Romania', 2.76),
 ('Croatia', 2.7300000000000004),
 ('Belarus', 2.66),
 ('Monaco', 2.59),
 ('Greece', 2.4299999999999997),
 ('Saint Pierre and Miquelon', 2.3000000000000007),
 ('Russia', 2.09),
 ('Estonia', 1.8900000000000006),
 ('Portugal', 1.75),
 ('Japan', 1.58),
 ('Italy', 1.4499999999999993),
 ('Bosnia and Herzegovina', 0.8800000000000008),
 ('Czech Republic', 0.7099999999999991),
 ('Moldova', 0.5899999999999999),
 ('Poland', 0.4499999999999993),
 ('Austria', 0.009999999999999787)]